trickle/trickle_server.go

```diff
@@ -0,0 +1,423 @@
+package trickle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// TODO sweep idle streams connections
+
+type StreamManager struct {
+	mutex   sync.RWMutex
+	streams map[string]*Stream
+}
+
+type Stream struct {
+	mutex       sync.RWMutex
+	segments    []*Segment
+	latestWrite int
+}
+
+type Segment struct {
+	idx    int
+	mutex  *sync.Mutex
+	cond   *sync.Cond
+	buffer *bytes.Buffer
+	closed bool
+}
+
+type SegmentSubscriber struct {
+	segment *Segment
+	readPos int
+}
+
+const maxSegmentsPerStream = 5
+
+const BaseServerPath = "/ai/live-video/"
+
+var FirstByteTimeout = errors.New("pending read timeout")
+
+func ConfigureServerWithMux(mux *http.ServeMux) {
+	/* TODO we probably want to configure the below
+	srv := &http.Server{
+		// say max segment size is 20 secs
+		// we can allow 2 * 20 secs given preconnects
+		ReadTimeout:  40 * time.Second,
+		WriteTimeout: 45 * time.Second,
+	}
+	*/
+
+	streamManager := &StreamManager{
+		streams: make(map[string]*Stream),
+	}
+	mux.HandleFunc("GET "+BaseServerPath+"{streamName}/{idx}", streamManager.handleGet)
+	mux.HandleFunc("POST "+BaseServerPath+"{streamName}/{idx}", streamManager.handlePost)
+	mux.HandleFunc("DELETE "+BaseServerPath+"{streamName}", streamManager.handleDelete)
+}
+
+func (sm *StreamManager) getStream(streamName string) (*Stream, bool) {
+	sm.mutex.RLock()
+	defer sm.mutex.RUnlock()
+	stream, exists := sm.streams[streamName]
+	return stream, exists
+}
+
+func (sm *StreamManager) getOrCreateStream(streamName string) *Stream {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	stream, exists := sm.streams[streamName]
+	if !exists {
+		stream = &Stream{
+			segments: make([]*Segment, 5),
+		}
+		sm.streams[streamName] = stream
+		slog.Info("Creating stream", "stream", streamName)
+	}
+	return stream
+}
+
+func (sm *StreamManager) clearAllStreams() {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	for _, stream := range sm.streams {
+		stream.clear()
+	}
+	sm.streams = make(map[string]*Stream)
+}
+
+func (s *Stream) clear() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	for _, segment := range s.segments {
+		segment.close()
+	}
+	s.segments = make([]*Segment, maxSegmentsPerStream)
+}
+
+func (sm *StreamManager) handleDelete(w http.ResponseWriter, r *http.Request) {
+	streamName := r.PathValue("streamName")
+	stream, exists := sm.getStream(streamName)
+	if !exists {
+		http.Error(w, "Invalid stream name", http.StatusBadRequest)
+		return
+	}
+
+	// TODO properly clear sessions once we have a good solution
```

Comment:
What is the proper solution for clearing the stream?

Reply:
Clearing the stream is not the problem, but session reuse was (eg, trying to start a new session with the same name). We can work around the issue for now by loosening up some of the constraints around sequence numbering. In practice I think this will be less of an issue as long as the O hands out a fresh ID for each session which it does [here](https://github.com/livepeer/go-livepeer/pull/3211/files/4db8c949d4edb8da4e1906dad4f1f36d3e82cbc3#diff-2d409e07b090aec3653ceddfa542b8809eea772f3f24920e8c285d1e22edd15bR256-R257).